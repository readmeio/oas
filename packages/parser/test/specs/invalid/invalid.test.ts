import { describe, it, expect, assert } from 'vitest';

import { SwaggerParser } from '../../../src/index.js';
import * as path from '../../utils/path.js';

describe("Invalid APIs (can't be parsed)", () => {
  it('not a Swagger API', async () => {
    try {
      await SwaggerParser.parse(path.rel('specs/invalid/not-swagger.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.contain('not-swagger.yaml is not a valid API definition');
    }
  });

  it('not a valid OpenAPI 3.1 definition', async () => {
    try {
      await SwaggerParser.parse(path.rel('specs/invalid/no-paths-or-webhooks.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.contain('no-paths-or-webhooks.yaml is not a valid OpenAPI definition');
    }
  });

  it('invalid Swagger version (1.2)', async () => {
    try {
      await SwaggerParser.dereference(path.rel('specs/invalid/old-version.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.equal('Unrecognized Swagger version: 1.2. Expected 2.0');
    }
  });

  it('invalid Swagger version (3.0)', async () => {
    try {
      await SwaggerParser.bundle(path.rel('specs/invalid/newer-version.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.equal('Unrecognized Swagger version: 3.0. Expected 2.0');
    }
  });

  it('numeric Swagger version (instead of a string)', async () => {
    try {
      await SwaggerParser.validate(path.rel('specs/invalid/numeric-version.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.equal('Swagger version number must be a string (e.g. "2.0") not a number.');
    }
  });

  it('numeric API version (instead of a string)', async () => {
    try {
      await SwaggerParser.validate(path.rel('specs/invalid/numeric-info-version.yaml'));
      assert.fail();
    } catch (err) {
      expect(err).to.be.an.instanceOf(SyntaxError);
      expect(err.message).to.equal('API version number must be a string (e.g. "1.0.0") not a number.');
    }
  });
});
