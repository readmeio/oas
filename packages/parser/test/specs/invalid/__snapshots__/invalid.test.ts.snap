// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Invalid APIs (can't be parsed) > invalid Swagger version (1.2) 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "ENUM must be equal to one of the allowed values
(2.0)

  1 | {
> 2 |   "swagger": "1.2",
    |              ^^^^^ Did you mean 2.0 here?
  3 |   "info": {
  4 |     "version": "1.0.0",
  5 |     "title": "old version",",
    },
  ],
  "specification": "Swagger",
  "valid": false,
  "warnings": [],
}
`;

exports[`Invalid APIs (can't be parsed) > invalid Swagger version (3.0) 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "ENUM must be equal to one of the allowed values
(2.0)

  1 | {
> 2 |   "swagger": "3.0",
    |              ^^^^^ Did you mean 2.0 here?
  3 |   "info": {
  4 |     "version": "1.0.0",
  5 |     "title": "newer version",",
    },
  ],
  "specification": "Swagger",
  "valid": false,
  "warnings": [],
}
`;

exports[`Invalid APIs (can't be parsed) > not a valid API definition 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "Supplied schema is not a valid API definition.",
    },
  ],
  "specification": null,
  "valid": false,
  "warnings": [],
}
`;

exports[`Invalid APIs (can't be parsed) > not a valid OpenAPI 3.1 definition 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "REQUIRED must have required property 'paths'

> 1 | {
    | ^ paths is missing here!
  2 |   "openapi": "3.1.0",
  3 |   "info": {
  4 |     "version": "1.0.0",",
    },
  ],
  "specification": "OpenAPI",
  "valid": false,
  "warnings": [],
}
`;

exports[`Invalid APIs (can't be parsed) > numeric API version (instead of a string) 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "TYPE must be string

  2 |   "swagger": "2.0",
  3 |   "info": {
> 4 |     "version": 1,
    |                ^ type must be string
  5 |     "title": "old version",
  6 |     "description": "This swagger spec is invalid because version 1.2 is not supported"
  7 |   },",
    },
  ],
  "specification": "Swagger",
  "valid": false,
  "warnings": [],
}
`;

exports[`Invalid APIs (can't be parsed) > numeric Swagger version (instead of a string) 1`] = `
{
  "additionalErrors": 0,
  "errors": [
    {
      "message": "TYPE must be string

  1 | {
> 2 |   "swagger": 2,
    |              ^ type must be string
  3 |   "info": {
  4 |     "version": "1.0.0",
  5 |     "title": "old version",",
    },
  ],
  "specification": "Swagger",
  "valid": false,
  "warnings": [],
}
`;
